File: Reference/src/Brontide.Reference.Experimental.Binding/Portable/PortableCbor.cs
using System.Collections.Immutable;
using System.Text;

namespace Brontide.Reference.Experimental.Binding.Portable;

/// <summary>A value in the deterministic CBOR core the portable representation permits.</summary>
/// <remarks>
/// The model is deliberately narrower than CBOR: there is no float, no indefinite length, no simple
/// value beyond true/false/null, and exactly one tag. A value outside the model cannot be
/// constructed, so the encoder cannot emit something the decoder would have to reject.
/// </remarks>
public abstract record CborItem;

public sealed record CborInteger(long Value) : CborItem;

public sealed record CborText(string Value) : CborItem;

public sealed record CborBytes(ReadOnlyMemory<byte> Value) : CborItem;

public sealed record CborBoolean(bool Value) : CborItem;

public sealed record CborNull : CborItem
{
    public static CborNull Instance { get; } = new();
}

/// <summary>The single allowlisted tag: an exact decimal fraction encoded as tag 4.</summary>
public sealed record CborDecimal(int Exponent, long Mantissa) : CborItem
{
    public const int MinExponent = -28;
    public const int MaxExponent = 28;

    /// <summary>
    /// Normalizes so that one value has exactly one encoding: a mantissa divisible by ten is scaled
    /// up, and a zero mantissa always carries exponent zero.
    /// </summary>
    public static CborDecimal Normalize(int exponent, long mantissa)
    {
        if (mantissa == 0)
        {
            return new CborDecimal(0, 0);
        }

        while (mantissa % 10 == 0 && exponent < MaxExponent)
        {
            mantissa /= 10;
            exponent++;
        }

        if (exponent is < MinExponent or > MaxExponent)
        {
            throw PortableFaultException.Malformed(
                "decimal-exponent",
                "A Decimal exponent outside -28..28 is not representable in the portable core.");
        }

        return new CborDecimal(exponent, mantissa);
    }
}

public sealed record CborArray(ImmutableArray<CborItem> Items) : CborItem
{
    public static CborArray Of(params CborItem[] items) => new([.. items]);
}

public sealed record CborMap(ImmutableArray<CborMapEntry> Entries) : CborItem
{
    public static CborMap Empty { get; } = new([]);

    public static CborMap Of(IEnumerable<(string Key, CborItem Value)> entries) =>
        new([.. entries.Select(entry => new CborMapEntry(entry.Key, entry.Value))]);

    public bool TryGet(string key, out CborItem value)
    {
        foreach (var entry in Entries)
        {
            if (string.Equals(entry.Key, key, StringComparison.Ordinal))
            {
                value = entry.Value;
                return true;
            }
        }

        value = CborNull.Instance;
        return false;
    }

    public bool Contains(string key) => TryGet(key, out _);
}

/// <summary>
/// One map entry. Every map key in this contract is a text string, so the key is modelled as one
/// rather than as a general item.
/// </summary>
public sealed record CborMapEntry(string Key, CborItem Value);

/// <summary>
/// The deterministic CBOR core encoder and decoder (RFC 8949 section 4.2.1), restricted to the
/// subset the portable representation declares.
/// </summary>
/// <remarks>
/// Map keys sort by the bytewise order of their complete encoding, not by ordinal string
/// comparison. The two orders disagree once key lengths cross an encoding-width boundary, so a
/// codec that reuses an ordinal comparer produces bytes this decoder rejects.
/// </remarks>
public static class PortableCbor
{
    private static readonly UTF8Encoding StrictUtf8 = new(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);

    public static byte[] Encode(CborItem item)
    {
        ArgumentNullException.ThrowIfNull(item);
        var buffer = new PortableByteBuffer();
        Write(buffer, item);
        return buffer.ToArray();
    }

    public static CborItem Decode(ReadOnlySpan<byte> body, PortableLimits limits)
    {
        ArgumentNullException.ThrowIfNull(limits);
        var position = 0;
        var item = Read(body, ref position, 1, limits);
        if (position != body.Length)
        {
            throw PortableFaultException.Malformed(
                "trailing-bytes",
                "One frame carries exactly one top-level item; trailing bytes are not permitted.");
        }

        return item;
    }

    /// <summary>Orders two complete key encodings by the deterministic bytewise rule.</summary>
    public static int CompareKeyEncodings(ReadOnlySpan<byte> left, ReadOnlySpan<byte> right)
    {
        var shared = Math.Min(left.Length, right.Length);
        for (var index = 0; index < shared; index++)
        {
            if (left[index] != right[index])
            {
                return left[index] < right[index] ? -1 : 1;
            }
        }

        return left.Length.CompareTo(right.Length);
    }

    // ---------------------------------------------------------------------
    // Encoding
    // ---------------------------------------------------------------------

    private static void Write(PortableByteBuffer buffer, CborItem item)
    {
        switch (item)
        {
            case CborInteger integer when integer.Value >= 0:
                WriteHead(buffer, 0, (ulong)integer.Value);
                break;
            case CborInteger integer:
                WriteHead(buffer, 1, unchecked((ulong)(-1 - integer.Value)));
                break;
            case CborBoolean boolean:
                buffer.Write(boolean.Value ? (byte)0xF5 : (byte)0xF4);
                break;
            case CborNull:
                buffer.Write((byte)0xF6);
                break;
            case CborText text:
            {
                var bytes = StrictUtf8.GetBytes(text.Value);
                WriteHead(buffer, 3, (ulong)bytes.Length);
                buffer.Write(bytes);
                break;
            }
            case CborBytes bytes:
                WriteHead(buffer, 2, (ulong)bytes.Value.Length);
                buffer.Write(bytes.Value.Span);
                break;
            case CborDecimal number:
            {
                var normalized = CborDecimal.Normalize(number.Exponent, number.Mantissa);
                WriteHead(buffer, 6, 4);
                WriteHead(buffer, 4, 2);
                Write(buffer, new CborInteger(normalized.Exponent));
                Write(buffer, new CborInteger(normalized.Mantissa));
                break;
            }
            case CborArray array:
                WriteHead(buffer, 4, (ulong)array.Items.Length);
                foreach (var element in array.Items)
                {
                    Write(buffer, element);
                }

                break;
            case CborMap map:
                WriteMap(buffer, map);
                break;
            default:
                throw PortableFaultException.Malformed(
                    "unencodable-item",
                    "The portable core cannot encode this item.");
        }
    }

    private static void WriteMap(PortableByteBuffer buffer, CborMap map)
    {
        var encoded = new List<(byte[] Key, byte[] Value)>(map.Entries.Length);
        var seen = new HashSet<string>(StringComparer.Ordinal);
        foreach (var entry in map.Entries)
        {
            if (!seen.Add(entry.Key))
            {
                throw PortableFaultException.Malformed(
                    "duplicate-key",
                    $"Map key '{entry.Key}' appears more than once.");
            }

            encoded.Add((Encode(new CborText(entry.Key)), Encode(entry.Value)));
        }

        encoded.Sort(static (left, right) => CompareKeyEncodings(left.Key, right.Key));
        WriteHead(buffer, 5, (ulong)encoded.Count);
        foreach (var entry in encoded)
        {
            buffer.Write(entry.Key);
            buffer.Write(entry.Value);
        }
    }

    private static void WriteHead(PortableByteBuffer buffer, int major, ulong argument)
    {
        var prefix = (byte)(major << 5);
        if (argument < 24)
        {
            buffer.Write((byte)(prefix | (byte)argument));
        }
        else if (argument <= byte.MaxValue)
        {
            buffer.Write((byte)(prefix | 24));
            buffer.Write((byte)argument);
        }
        else if (argument <= ushort.MaxValue)
        {
            buffer.Write((byte)(prefix | 25));
            WriteBigEndian(buffer, argument, 2);
        }
        else if (argument <= uint.MaxValue)
        {
            buffer.Write((byte)(prefix | 26));
            WriteBigEndian(buffer, argument, 4);
        }
        else
        {
            buffer.Write((byte)(prefix | 27));
            WriteBigEndian(buffer, argument, 8);
        }
    }

    private static void WriteBigEndian(PortableByteBuffer buffer, ulong argument, int width)
    {
        for (var shift = (width - 1) * 8; shift >= 0; shift -= 8)
        {
            buffer.Write((byte)((argument >> shift) & 0xFF));
        }
    }

    // ---------------------------------------------------------------------
    // Decoding
    // ---------------------------------------------------------------------

    private static CborItem Read(ReadOnlySpan<byte> body, ref int position, int depth, PortableLimits limits)
    {
        if (depth > limits.MaxNestingDepth)
        {
            throw PortableFaultException.LimitExceeded(
                "nesting-depth",
                $"Decoding stopped at the declared nesting depth of {limits.MaxNestingDepth}.");
        }

        var head = Take(body, ref position, 1)[0];
        var major = head >> 5;
        var additional = head & 0x1F;

        if (major == 7)
        {
            return additional switch
            {
                20 => new CborBoolean(false),
                21 => new CborBoolean(true),
                22 => CborNull.Instance,
                _ => throw PortableFaultException.Malformed(
                    "simple-value",
                    "Only the simple values 20, 21, and 22 belong to the deterministic core; floats are excluded.")
            };
        }

        var argument = ReadArgument(body, ref position, additional);
        switch (major)
        {
            case 0:
                if (argument > long.MaxValue)
                {
                    throw PortableFaultException.Malformed(
                        "integer-domain",
                        "An unsigned integer outside the Integer.Signed64 domain is not portable.");
                }

                return new CborInteger((long)argument);
            case 1:
                if (argument > long.MaxValue)
                {
                    throw PortableFaultException.Malformed(
                        "integer-domain",
                        "A negative integer outside the Integer.Signed64 domain is not portable.");
                }

                return new CborInteger(unchecked(-1L - (long)argument));
            case 2:
            {
                var length = RequireLength(argument, limits.MaxByteStringBytes, "byte-string");
                return new CborBytes(Take(body, ref position, length).ToArray());
            }
            case 3:
            {
                var length = RequireLength(argument, limits.MaxTextBytes, "text-string");
                var bytes = Take(body, ref position, length);
                try
                {
                    return new CborText(StrictUtf8.GetString(bytes));
                }
                catch (DecoderFallbackException)
                {
                    throw PortableFaultException.Malformed("text-encoding", "A text string is not well-formed UTF-8.");
                }
            }
            case 4:
            {
                var count = RequireLength(argument, limits.MaxSequenceItems, "array");
                var items = ImmutableArray.CreateBuilder<CborItem>(count);
                for (var index = 0; index < count; index++)
                {
                    items.Add(Read(body, ref position, depth + 1, limits));
                }

                return new CborArray(items.MoveToImmutable());
            }
            case 5:
                return ReadMap(body, ref position, depth, limits, argument);
            case 6:
                return ReadTag(body, ref position, depth, limits, argument);
            default:
                throw PortableFaultException.Malformed("major-type", "The item uses a major type outside the portable core.");
        }
    }

    private static CborMap ReadMap(
        ReadOnlySpan<byte> body,
        ref int position,
        int depth,
        PortableLimits limits,
        ulong argument)
    {
        var count = RequireLength(argument, limits.MaxRecordFields, "map");
        var entries = ImmutableArray.CreateBuilder<CborMapEntry>(count);
        var previousKeyStart = -1;
        var previousKeyEnd = -1;
        for (var index = 0; index < count; index++)
        {
            var keyStart = position;
            var key = Read(body, ref position, depth + 1, limits) as CborText ??
                throw PortableFaultException.Malformed("map-key-kind", "Every portable map key is a text string.");
            var keyEnd = position;

            if (previousKeyStart >= 0)
            {
                var order = CompareKeyEncodings(
                    body[previousKeyStart..previousKeyEnd],
                    body[keyStart..keyEnd]);
                if (order == 0)
                {
                    throw PortableFaultException.Malformed(
                        "duplicate-key",
                        $"Map key '{key.Value}' appears more than once.");
                }

                if (order > 0)
                {
                    throw PortableFaultException.Malformed(
                        "map-key-order",
                        $"Map key '{key.Value}' breaks ascending deterministic key order.");
                }
            }

            previousKeyStart = keyStart;
            previousKeyEnd = keyEnd;
            entries.Add(new CborMapEntry(key.Value, Read(body, ref position, depth + 1, limits)));
        }

        return new CborMap(entries.MoveToImmutable());
    }

    private static CborDecimal ReadTag(
        ReadOnlySpan<byte> body,
        ref int position,
        int depth,
        PortableLimits limits,
        ulong tag)
    {
        if (tag != 4)
        {
            throw PortableFaultException.Malformed("tag-allowlist", $"Tag {tag} is outside the allowlist, which contains only tag 4.");
        }

        if (Read(body, ref position, depth + 1, limits) is not CborArray array ||
            array.Items.Length != 2 ||
            array.Items[0] is not CborInteger exponent ||
            array.Items[1] is not CborInteger mantissa)
        {
            throw PortableFaultException.Malformed(
                "decimal-body",
                "Tag 4 carries an array of exactly two integers: [exponent, mantissa].");
        }

        if (exponent.Value is < CborDecimal.MinExponent or > CborDecimal.MaxExponent)
        {
            throw PortableFaultException.Malformed("decimal-exponent", "A Decimal exponent must lie in -28..28.");
        }

        if (mantissa.Value == 0)
        {
            if (exponent.Value != 0)
            {
                throw PortableFaultException.Malformed(
                    "decimal-canonical",
                    "A zero mantissa is canonically encoded with exponent 0.");
            }
        }
        else if (mantissa.Value % 10 == 0)
        {
            throw PortableFaultException.Malformed(
                "decimal-canonical",
                "A Decimal mantissa is normalized so that it is not divisible by ten.");
        }

        return new CborDecimal((int)exponent.Value, mantissa.Value);
    }

    private static ulong ReadArgument(ReadOnlySpan<byte> body, ref int position, int additional)
    {
        switch (additional)
        {
            case < 24:
                return (ulong)additional;
            case 24:
            {
                var value = (ulong)Take(body, ref position, 1)[0];
                return RequireShortest(value, 24);
            }
            case 25:
                return RequireShortest(ReadBigEndian(body, ref position, 2), byte.MaxValue + 1UL);
            case 26:
                return RequireShortest(ReadBigEndian(body, ref position, 4), ushort.MaxValue + 1UL);
            case 27:
                return RequireShortest(ReadBigEndian(body, ref position, 8), uint.MaxValue + 1UL);
            case 31:
                throw PortableFaultException.Malformed(
                    "indefinite-length",
                    "Indefinite-length items are excluded from the deterministic core.");
            default:
                throw PortableFaultException.Malformed("reserved-argument", "The argument uses a reserved additional-information value.");
        }
    }

    private static ulong ReadBigEndian(ReadOnlySpan<byte> body, ref int position, int width)
    {
        var bytes = Take(body, ref position, width);
        var value = 0UL;
        foreach (var current in bytes)
        {
            value = (value << 8) | current;
        }

        return value;
    }

    private static ulong RequireShortest(ulong value, ulong minimum) =>
        value >= minimum
            ? value
            : throw PortableFaultException.Malformed(
                "non-shortest-argument",
                "An argument must use the shortest form that represents it exactly.");

    private static int RequireLength(ulong argument, int bound, string what)
    {
        if (argument > (ulong)bound)
        {
            throw PortableFaultException.LimitExceeded(
                $"{what}-bound",
                $"A {what} of {argument} exceeds the declared bound of {bound}.");
        }

        return (int)argument;
    }

    private static ReadOnlySpan<byte> Take(ReadOnlySpan<byte> body, ref int position, int count)
    {
        if (count < 0 || position + count > body.Length)
        {
            throw PortableFaultException.Malformed("truncated-item", "The frame body ends inside an item.");
        }

        var slice = body.Slice(position, count);
        position += count;
        return slice;
    }
}

/// <summary>A minimal growable byte sink; the codec writes forward only and never seeks.</summary>
internal sealed class PortableByteBuffer
{
    private byte[] _buffer = new byte[256];
    private int _count;

    public void Write(byte value)
    {
        EnsureCapacity(1);
        _buffer[_count++] = value;
    }

    public void Write(ReadOnlySpan<byte> values)
    {
        EnsureCapacity(values.Length);
        values.CopyTo(_buffer.AsSpan(_count));
        _count += values.Length;
    }

    public byte[] ToArray() => _buffer.AsSpan(0, _count).ToArray();

    private void EnsureCapacity(int additional)
    {
        if (_count + additional <= _buffer.Length)
        {
            return;
        }

        var capacity = _buffer.Length;
        while (capacity < _count + additional)
        {
            capacity *= 2;
        }

        Array.Resize(ref _buffer, capacity);
    }
}

File: Minimal/src/Brontide.Minimal.Binding/Portable/PortableCbor.fs
namespace Brontide.Minimal.Binding.Portable

open System
open System.Text

/// A value in the deterministic CBOR core the portable representation permits.
///
/// The union is deliberately narrower than CBOR: no float, no indefinite length, no simple value
/// beyond true/false/null, and exactly one tag. A value outside the union cannot be constructed, so
/// the encoder cannot emit something the decoder would have to refuse.
type CborItem =
    | CborInteger of int64
    | CborText of string
    | CborBytes of byte array
    | CborBoolean of bool
    | CborNull
    /// The single allowlisted tag: an exact decimal fraction encoded as tag 4.
    | CborDecimal of exponent: int * mantissa: int64
    | CborArray of CborItem list
    | CborMap of (string * CborItem) list

[<RequireQualifiedAccess>]
module CborDecimal =
    [<Literal>]
    let MinExponent = -28

    [<Literal>]
    let MaxExponent = 28

    /// Normalizes so that one value has exactly one encoding: a mantissa divisible by ten is scaled
    /// up, and a zero mantissa always carries exponent zero.
    let normalize exponent mantissa : PortableResult<int * int64> =
        if mantissa = 0L then
            Ok(0, 0L)
        else
            let rec scale exponent mantissa =
                if mantissa % 10L = 0L && exponent < MaxExponent then
                    scale (exponent + 1) (mantissa / 10L)
                else
                    exponent, mantissa

            let scaledExponent, scaledMantissa = scale exponent mantissa

            if scaledExponent < MinExponent || scaledExponent > MaxExponent then
                malformed "decimal-exponent" "A Decimal exponent outside -28..28 is not representable in the portable core."
            else
                Ok(scaledExponent, scaledMantissa)

/// The deterministic CBOR core encoder and decoder (RFC 8949 section 4.2.1), restricted to the
/// subset the portable representation declares.
///
/// Map keys sort by the bytewise order of their complete encoding, not by ordinal string
/// comparison. The two orders disagree once key lengths cross an encoding-width boundary, which is
/// why the baseline JSON codec's comparer cannot be reused here.
[<RequireQualifiedAccess>]
module PortableCbor =

    let private strictUtf8 = UTF8Encoding(false, true)

    /// Orders two complete key encodings by the deterministic bytewise rule.
    let compareKeyEncodings (left: byte array) (leftStart: int) (leftLength: int) (right: byte array) (rightStart: int) (rightLength: int) =
        let shared = min leftLength rightLength

        let rec walk index =
            if index = shared then compare leftLength rightLength
            else
                let l = left.[leftStart + index]
                let r = right.[rightStart + index]
                if l = r then walk (index + 1)
                elif l < r then -1
                else 1

        walk 0

    // -- encoding ---------------------------------------------------------

    let private writeHead (buffer: ResizeArray<byte>) (major: int) (argument: uint64) =
        let prefix = byte (major <<< 5)

        let writeBigEndian width =
            for shift in [ (width - 1) * 8 .. -8 .. 0 ] do
                buffer.Add(byte ((argument >>> shift) &&& 0xFFUL))

        if argument < 24UL then
            buffer.Add(prefix ||| byte argument)
        elif argument <= uint64 Byte.MaxValue then
            buffer.Add(prefix ||| 24uy)
            buffer.Add(byte argument)
        elif argument <= uint64 UInt16.MaxValue then
            buffer.Add(prefix ||| 25uy)
            writeBigEndian 2
        elif argument <= uint64 UInt32.MaxValue then
            buffer.Add(prefix ||| 26uy)
            writeBigEndian 4
        else
            buffer.Add(prefix ||| 27uy)
            writeBigEndian 8

    let rec private write (buffer: ResizeArray<byte>) (item: CborItem) : PortableResult<unit> =
        match item with
        | CborInteger value when value >= 0L ->
            writeHead buffer 0 (uint64 value)
            Ok()
        | CborInteger value ->
            writeHead buffer 1 (uint64 (-1L - value))
            Ok()
        | CborBoolean value ->
            buffer.Add(if value then 0xF5uy else 0xF4uy)
            Ok()
        | CborNull ->
            buffer.Add 0xF6uy
            Ok()
        | CborText value ->
            let bytes = strictUtf8.GetBytes value
            writeHead buffer 3 (uint64 bytes.Length)
            buffer.AddRange bytes
            Ok()
        | CborBytes value ->
            writeHead buffer 2 (uint64 value.Length)
            buffer.AddRange value
            Ok()
        | CborDecimal(exponent, mantissa) ->
            CborDecimal.normalize exponent mantissa
            |> Result.bind (fun (normalizedExponent, normalizedMantissa) ->
                writeHead buffer 6 4UL
                writeHead buffer 4 2UL

                portable {
                    do! write buffer (CborInteger(int64 normalizedExponent))
                    do! write buffer (CborInteger normalizedMantissa)
                })
        | CborArray items ->
            writeHead buffer 4 (uint64 (List.length items))
            items |> iterate (write buffer)
        | CborMap entries -> writeMap buffer entries

    and private writeMap (buffer: ResizeArray<byte>) (entries: (string * CborItem) list) : PortableResult<unit> =
        let duplicate =
            entries
            |> List.countBy fst
            |> List.tryFind (fun (_, count) -> count > 1)

        match duplicate with
        | Some(key, _) -> malformed "duplicate-key" $"Map key '{key}' appears more than once."
        | None ->
            let encodeEntry (key, value) =
                portable {
                    let keyBuffer = ResizeArray<byte>()
                    do! write keyBuffer (CborText key)
                    let valueBuffer = ResizeArray<byte>()
                    do! write valueBuffer value
                    return keyBuffer.ToArray(), valueBuffer.ToArray()
                }

            entries
            |> traverse encodeEntry
            |> Result.map (fun encoded ->
                let sorted =
                    encoded
                    |> List.sortWith (fun (leftKey: byte array, _) (rightKey: byte array, _) ->
                        compareKeyEncodings leftKey 0 leftKey.Length rightKey 0 rightKey.Length)

                writeHead buffer 5 (uint64 (List.length sorted))

                for key, value in sorted do
                    buffer.AddRange key
                    buffer.AddRange value)

    let encode (item: CborItem) : PortableResult<byte array> =
        let buffer = ResizeArray<byte>()
        write buffer item |> Result.map (fun () -> buffer.ToArray())

    // -- decoding ---------------------------------------------------------

    let private take (body: byte array) (position: int) (count: int) : PortableResult<int> =
        if count < 0 || position + count > body.Length then
            malformed "truncated-item" "The frame body ends inside an item."
        else
            Ok(position + count)

    let private requireShortest (value: uint64) (minimum: uint64) =
        if value >= minimum then
            Ok value
        else
            malformed "non-shortest-argument" "An argument must use the shortest form that represents it exactly."

    let private readBigEndian (body: byte array) (position: int) (width: int) : PortableResult<uint64 * int> =
        take body position width
        |> Result.map (fun next ->
            let mutable value = 0UL

            for index in position .. next - 1 do
                value <- (value <<< 8) ||| uint64 body.[index]

            value, next)

    let private readArgument (body: byte array) (position: int) (additional: int) : PortableResult<uint64 * int> =
        if additional < 24 then
            Ok(uint64 additional, position)
        elif additional = 24 then
            take body position 1
            |> Result.bind (fun next ->
                requireShortest (uint64 body.[position]) 24UL
                |> Result.map (fun value -> value, next))
        elif additional = 25 then
            readBigEndian body position 2
            |> Result.bind (fun (value, next) ->
                requireShortest value (uint64 Byte.MaxValue + 1UL) |> Result.map (fun value -> value, next))
        elif additional = 26 then
            readBigEndian body position 4
            |> Result.bind (fun (value, next) ->
                requireShortest value (uint64 UInt16.MaxValue + 1UL) |> Result.map (fun value -> value, next))
        elif additional = 27 then
            readBigEndian body position 8
            |> Result.bind (fun (value, next) ->
                requireShortest value (uint64 UInt32.MaxValue + 1UL) |> Result.map (fun value -> value, next))
        elif additional = 31 then
            malformed "indefinite-length" "Indefinite-length items are excluded from the deterministic core."
        else
            malformed "reserved-argument" "The argument uses a reserved additional-information value."

    let private requireLength (argument: uint64) (bound: int) (what: string) : PortableResult<int> =
        if argument > uint64 bound then
            limitExceeded $"{what}-bound" $"A {what} of {argument} exceeds the declared bound of {bound}."
        else
            Ok(int argument)

    let rec private read (body: byte array) (position: int) (depth: int) (limits: PortableLimits) : PortableResult<CborItem * int> =
        if depth > limits.MaxNestingDepth then
            limitExceeded "nesting-depth" $"Decoding stopped at the declared nesting depth of {limits.MaxNestingDepth}."
        else
            take body position 1
            |> Result.bind (fun afterHead ->
                let head = body.[position]
                let major = int (head >>> 5)
                let additional = int (head &&& 0x1Fuy)

                if major = 7 then
                    match additional with
                    | 20 -> Ok(CborBoolean false, afterHead)
                    | 21 -> Ok(CborBoolean true, afterHead)
                    | 22 -> Ok(CborNull, afterHead)
                    | _ ->
                        malformed
                            "simple-value"
                            "Only the simple values 20, 21, and 22 belong to the deterministic core; floats are excluded."
                else
                    readArgument body afterHead additional
                    |> Result.bind (fun (argument, afterArgument) ->
                        match major with
                        | 0 ->
                            if argument > uint64 Int64.MaxValue then
                                malformed "integer-domain" "An unsigned integer outside the Integer.Signed64 domain is not portable."
                            else
                                Ok(CborInteger(int64 argument), afterArgument)
                        | 1 ->
                            if argument > uint64 Int64.MaxValue then
                                malformed "integer-domain" "A negative integer outside the Integer.Signed64 domain is not portable."
                            else
                                Ok(CborInteger(-1L - int64 argument), afterArgument)
                        | 2 ->
                            requireLength argument limits.MaxByteStringBytes "byte-string"
                            |> Result.bind (fun length ->
                                take body afterArgument length
                                |> Result.map (fun next -> CborBytes(Array.sub body afterArgument length), next))
                        | 3 ->
                            requireLength argument limits.MaxTextBytes "text-string"
                            |> Result.bind (fun length ->
                                take body afterArgument length
                                |> Result.bind (fun next ->
                                    try
                                        Ok(CborText(strictUtf8.GetString(body, afterArgument, length)), next)
                                    with :? DecoderFallbackException ->
                                        malformed "text-encoding" "A text string is not well-formed UTF-8."))
                        | 4 ->
                            requireLength argument limits.MaxSequenceItems "array"
                            |> Result.bind (fun count -> readArray body afterArgument depth limits count)
                        | 5 ->
                            requireLength argument limits.MaxRecordFields "map"
                            |> Result.bind (fun count -> readMap body afterArgument depth limits count)
                        | 6 -> readTag body afterArgument depth limits argument
                        | _ -> malformed "major-type" "The item uses a major type outside the portable core."))

    and private readArray body position depth limits count : PortableResult<CborItem * int> =
        let rec walk index accumulated cursor =
            if index = count then
                Ok(CborArray(List.rev accumulated), cursor)
            else
                match read body cursor (depth + 1) limits with
                | Ok(item, next) -> walk (index + 1) (item :: accumulated) next
                | Error error -> Error error

        walk 0 [] position

    and private readMap body position depth limits count : PortableResult<CborItem * int> =
        let rec walk index accumulated cursor previousKey =
            if index = count then
                Ok(CborMap(List.rev accumulated), cursor)
            else
                let keyStart = cursor

                match read body cursor (depth + 1) limits with
                | Error error -> Error error
                | Ok(keyItem, afterKey) ->
                    match keyItem with
                    | CborText key ->
                        let ordering =
                            match previousKey with
                            | Some(previousStart, previousLength) ->
                                compareKeyEncodings body previousStart previousLength body keyStart (afterKey - keyStart)
                            | None -> -1

                        if ordering = 0 then
                            malformed "duplicate-key" $"Map key '{key}' appears more than once."
                        elif ordering > 0 then
                            malformed "map-key-order" $"Map key '{key}' breaks ascending deterministic key order."
                        else
                            match read body afterKey (depth + 1) limits with
                            | Error error -> Error error
                            | Ok(value, afterValue) ->
                                walk (index + 1) ((key, value) :: accumulated) afterValue (Some(keyStart, afterKey - keyStart))
                    | _ -> malformed "map-key-kind" "Every portable map key is a text string."

        walk 0 [] position None

    and private readTag body position depth limits tag : PortableResult<CborItem * int> =
        if tag <> 4UL then
            malformed "tag-allowlist" $"Tag {tag} is outside the allowlist, which contains only tag 4."
        else
            read body position (depth + 1) limits
            |> Result.bind (fun (item, next) ->
                match item with
                | CborArray [ CborInteger exponent; CborInteger mantissa ] ->
                    if exponent < int64 CborDecimal.MinExponent || exponent > int64 CborDecimal.MaxExponent then
                        malformed "decimal-exponent" "A Decimal exponent must lie in -28..28."
                    elif mantissa = 0L && exponent <> 0L then
                        malformed "decimal-canonical" "A zero mantissa is canonically encoded with exponent 0."
                    elif mantissa <> 0L && mantissa % 10L = 0L then
                        malformed "decimal-canonical" "A Decimal mantissa is normalized so that it is not divisible by ten."
                    else
                        Ok(CborDecimal(int exponent, mantissa), next)
                | _ -> malformed "decimal-body" "Tag 4 carries an array of exactly two integers: [exponent, mantissa].")

    let decode (body: byte array) (limits: PortableLimits) : PortableResult<CborItem> =
        read body 0 1 limits
        |> Result.bind (fun (item, position) ->
            if position <> body.Length then
                malformed "trailing-bytes" "One frame carries exactly one top-level item; trailing bytes are not permitted."
            else
                Ok item)

/// Strict accessors over a decoded control item. Every miss is a portable category rather than a
/// parser failure, so no foreign runtime failure escapes the decode boundary.
[<RequireQualifiedAccess>]
module CborAccess =

    let requireMap (item: CborItem) (what: string) : PortableResult<(string * CborItem) list> =
        match item with
        | CborMap entries -> Ok entries
        | _ -> malformed $"{what}-kind" $"'{what}' must be a map."

    let requireArray (item: CborItem) (what: string) : PortableResult<CborItem list> =
        match item with
        | CborArray items -> Ok items
        | _ -> malformed $"{what}-kind" $"'{what}' must be an array."

    let tryField (entries: (string * CborItem) list) (name: string) =
        entries |> List.tryPick (fun (key, value) -> if key = name then Some value else None)

    let contains entries name = (tryField entries name).IsSome

    let field entries name : PortableResult<CborItem> =
        match tryField entries name with
        | Some value -> Ok value
        | None -> malformed $"{name}-absent" $"The required field '{name}' is absent."

    let text entries name : PortableResult<string> =
        field entries name
        |> Result.bind (fun item ->
            match item with
            | CborText value -> Ok value
            | _ -> malformed $"{name}-kind" $"'{name}' must be a text string.")

    let integer entries name : PortableResult<int64> =
        field entries name
        |> Result.bind (fun item ->
            match item with
            | CborInteger value -> Ok value
            | _ -> malformed $"{name}-kind" $"'{name}' must be an integer.")

    let int32 entries name : PortableResult<int> =
        integer entries name
        |> Result.bind (fun value ->
            if value >= int64 Int32.MinValue && value <= int64 Int32.MaxValue then
                Ok(int value)
            else
                malformed $"{name}-range" $"'{name}' is outside the 32-bit range.")

    let boolean entries name : PortableResult<bool> =
        field entries name
        |> Result.bind (fun item ->
            match item with
            | CborBoolean value -> Ok value
            | _ -> malformed $"{name}-kind" $"'{name}' must be a boolean.")

    let map entries name : PortableResult<(string * CborItem) list> =
        field entries name |> Result.bind (fun item -> requireMap item name)

    let array entries name : PortableResult<CborItem list> =
        field entries name |> Result.bind (fun item -> requireArray item name)

    let arrayOf entries name (read: CborItem -> PortableResult<'T>) : PortableResult<'T list> =
        array entries name |> Result.bind (traverse read)

    let optionalText entries name : PortableResult<string option> =
        match tryField entries name with
        | None -> Ok None
        | Some(CborText value) -> Ok(Some value)
        | Some _ -> malformed $"{name}-kind" $"'{name}' must be a text string."

    let requireText (item: CborItem) (what: string) : PortableResult<string> =
        match item with
        | CborText value -> Ok value
        | _ -> malformed $"{what}-kind" $"'{what}' must be a text string."

    /// Refuses any field the caller did not declare, per the reject-unknown-field policy.
    let requireDeclaredFields (entries: (string * CborItem) list) (what: string) (declared: string list) =
        entries
        |> iterate (fun (key, _) ->
            if List.contains key declared then
                Ok()
            else
                malformed $"{what}-unknown-field" $"'{what}' declares no field named '{key}'.")

    let encodeCanonical (canonical: PortableCanonical) =
        CborMap
            [ "name", CborText(PortableCanonical.name canonical)
              "version", CborInteger(int64 (PortableCanonical.version canonical)) ]

    let readCanonical (item: CborItem) (what: string) : PortableResult<PortableCanonical> =
        portable {
            let! entries = requireMap item what
            do! requireDeclaredFields entries what [ "name"; "version" ]
            let! name = text entries "name"
            let! version = int32 entries "version"
            return! PortableCanonical.tryCreate name version
        }
